File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys

sys.setrecursionlimit(1000000)

N = int(sys.stdin.readline())
Nlist =[[] for _ in range(N+1)]

for i in range(N-1):
    a,b=list(map(int,sys.stdin.readline().split()))
    Nlist[a].append(b)
    Nlist[b].append(a)


visited = [0] * (N+1)
#0이 안방문한거

def DFS(startindex) :
    for i in Nlist[startindex]:
        if visited[i] == 0 :
            visited[i] = startindex
            DFS(i)
DFS(1)

for i in range(2,len(visited)) :
    print(visited[i])
    

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    static int N ; 
    static List<List<Integer>> Narr = new ArrayList<>();
    static int [] whoisdaddy ;
    static boolean[] visited ; 
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());

        for (int i = 0 ; i<=N ; i++)
            {
                List<Integer> newlist=  new ArrayList<>();
                Narr.add(newlist);
            }

        whoisdaddy = new int [N+1];
        visited= new boolean[N+1];
        
        for (int i = 0 ; i<N-1;  i++)
            {
                String [] oneline = br.readLine().split(" ");
                int x = Integer.parseInt(oneline[0]);
                int y = Integer.parseInt(oneline[1]);

                Narr.get(x).add(y);
                Narr.get(y).add(x);
            }

        DFS(1,-1);

        for (int i = 2; i<=N ; i++)
            {
                bw.write(String.valueOf(whoisdaddy[i]) + "\n");
            }
        bw.close();
    }

    public static void DFS(int now, int father)
    {
        visited[now] = true;
        whoisdaddy[now] = father;

        for (Integer val : Narr.get(now))
            {
                if(!visited[val])
                {
                    DFS(val,now);
                }
            }
    }
}